capa/features/extractors/vmray/__init__.py

```diff
@@ -128,41 +147,42 @@ def _compute_sections(self):
             for elffile_section in self.sample_file_static_data.elf.sections:
                 self.sections[elffile_section.header.sh_addr] = elffile_section.header.sh_name
 
-    def _compute_process_ids(self):
+    def _compute_monitor_processes(self):
         for process in self.sv2.processes.values():
-            # we expect VMRay's monitor IDs to be unique, but OS PIDs may be reused
-            assert process.monitor_id not in self.os_pid_by_monitor_id.keys()
-            self.os_pid_by_monitor_id[process.monitor_id] = process.os_pid
+            # we expect monitor IDs to be unique
+            assert process.monitor_id not in self.monitor_processes.keys()
 
-        # not all processes may get an ID, get missing data from flog.xml, see #2394
+            ppid: int = (
+                self.sv2.processes[process.ref_parent_process.path[1]].os_pid if process.ref_parent_process else 0
+            )
+            self.monitor_processes[process.monitor_id] = VMRayMonitorProcess(
+                process.os_pid, ppid, process.monitor_id, process.image_name
+            )
+
+        # not all processes are recorded in SummaryV2.json, get missing data from flog.xml, see #2394
         for monitor_process in self.flog.analysis.monitor_processes:
-            if monitor_process.process_id in self.os_pid_by_monitor_id:
-                assert self.os_pid_by_monitor_id[monitor_process.process_id] == monitor_process.os_pid
-            else:
-                self.os_pid_by_monitor_id[monitor_process.process_id] = monitor_process.os_pid
-
-    def _compute_process_threads(self):
-        # logs/flog.xml appears to be the only file that contains thread-related data
-        # so we use it here to map processes to threads
-        for function_call in self.flog.analysis.function_calls:
-            pid: int = self.get_process_os_pid(function_call.process_id)  # flog.xml uses process monitor ID, not OS PID
-            tid: int = function_call.thread_id
+            if monitor_process.process_id not in self.monitor_processes.keys():
+                self.monitor_processes[monitor_process.process_id] = VMRayMonitorProcess(
+                    monitor_process.os_pid,
+                    monitor_process.os_parent_pid,
+                    monitor_process.process_id,
+                    monitor_process.image_name,
+                )
+
+    def _compute_monitor_threads(self):
+        for monitor_thread in self.flog.analysis.monitor_threads:
+            # we expect monitor IDs to be unique
+            assert monitor_thread.thread_id not in self.monitor_threads.keys()
+
+            self.monitor_threads[monitor_thread.thread_id] = VMRayMonitorThread(
+                monitor_thread.os_tid, monitor_thread.thread_id, monitor_thread.process_id
+            )
 
-            assert isinstance(pid, int)
-            assert isinstance(tid, int)
+            # we expect 1 monitor thread per monitor process
```

Comment:
this comment confuses me (only 1 thread per process?) and the following 2 lines appear unrelated?

Reply:
Ah yes I can see how this is comment is confusing. I intended to communicate that we expect each monitor thread ID that is associated with each monitor process ID to be unique, e.g. the monitor thread ID 10 should not be associated with monitor process ID 1 more than once. This is because it appears that monitor process/thread IDs are incrementally unique for each process/thread monitored by VMRay. I'll attempt to clarify this better with an updated comment.